### ISSUE TITLE
feature/RFC: assign array from subshell

### ISSUE THREAD

**Issue by LukeShu** [koalaman/shellcheck#1020]
#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

When Bash scripting, sometimes we want to get an array out of a subshell the way we do a string with command substitution.  Doing this robustly turns out to be surprisingly difficult.

That is: how do we robustly do something like the following?

    outer=($(do_stuff && echo "${inner[*]}"))
    # NB: shellcheck 0.4.6 doesn't complain about the above,
    # while master would have us change it to
    read -r -a outer < <(do_stuff && echo "${inner[*]}")
    # which is still not quite right

Note that we *do* care about the return code; and would like to be able to handle errors from `do_stuff`.  Also note that the array members may contain arbitrary newlines, whitespace, or glob characters.

I have identified 2 methods for robustly handling this--unfortunately, shellcheck complains about both!  Obviously, inserting a shellcheck directive (or bogus assignment, in the case of SC2154) is an option, but it's a hassle to have to insert a directive every time we don an assignment of this type, especially since less robust methods may come back clean!

```
str=$(do_stuff && printf '%q ' "${inner[@]}")
# handle $?
outer=() # XXX: having to insert this line to avoid SC2154 is silly
eval "outer=($str)"
```
or
```
array_print() {
	local IFS=' '
	# shellcheck disable=SC1003
	set -- "${@//'\'/'\\'}"
	set -- "${@//' '/'\ '}"
	printf '%sx' "$*"
}

str=$(do_stuff && array_printprint_array "${inner[@]}")
# handle $?
# shellcheck disable=SC2162 # XXX: having to insert this for every such assignment is silly
IFS=' ' read -a outer -d '' < <(printf '%s' "${str%x}")
```
Now, my first reaction is to shove the problematic code into a function, so a shellcheck directive only needs to be used once, where the function is declared.  However, AFAICT, `getModifiedVariableCommand` can't be made to recognize user-defined functions.  (I even opened #1017 hoping that it could be made to recognize a function containing the second solution for what it is.)

I get that I can use the directives on each statement.  I get that I could globally disable the error code.   But I feel that right now, shellcheck is pushing toward *less robust* solutions, which is the Wrong Thing.

I think my desired outcome is that shellcheck gains a directive that identifies a user-defined function as a "modified variable command" that sets the argument named in the n-th variable.  Maybe something like

```
# shellcheck modify=$1
set_array() {
	eval "$1=($2)"
}
```

However, because I may be falling down the XY problem when I say that (just as I did when I opened #1017), that's not what I made the issue about.  If there is another, better solution, I would absolutely welcome it.

**Comment by mehrdadn** [comment]
I have similar problems. I'm just trying to do

  IFS=":" GLOBIGNORE="*" command eval 'basenames=($(eval basename --multiple -- "${names[@]}"))'

but then it warns me that:

- `names` is unused  
- Expressions don't expand in single quotes, use double quotes for that.  
- `basenames` is unassigned

...all of which are false positives. It'd be nice if it tried to work out the simple cases and realize what is going on.
